File: 02/main.js
const text = require('fs').readFileSync('input.txt').toString()
const commands = text.split('\n').map((item) => item.split(' '))

let depth = 0
let horiz = 0
let aim = 0

for (let [ com, val ] of commands) {
    if (com === 'down') {
        aim += parseInt(val)
    } else if (com === 'up') {
        aim -= parseInt(val)
    } else if (com === 'forward') {
        horiz += parseInt(val)
        depth += aim * parseInt(val)
    }
}

console.log(depth * horiz)

File: 02/main.hs
parseCommand line = let [com, val] = words line in (com, read val :: Int)

processCommand (depth, horiz, aim) (com, val) = case com of
  "down" -> (depth, horiz, aim + val)
  "up" -> (depth, horiz, aim - val)
  "forward" -> (depth + (aim * val), horiz + val, aim)

main = interact $ \input ->
  let (depth, horiz, _) = foldl processCommand (0, 0, 0) $ map parseCommand $ lines input in show $ depth * horiz